### PR TITLE
Add dependency checksums to quarkus caching extension

### PR DESCRIPTION
**Pull request by jprinet**
### Context
This change is the last one before v1.0.
The Quarkus Maven plugin dynamically adds some dependencies to the `classpath`, which must be declared as additional inputs.
[This change](https://github.com/quarkusio/quarkus/pull/37622) on Quarkus enables the creation of a file containing this list of dependencies.
This file is the one to add as goal input.
This dynamic `classpath` change is also happening for the tests (`surefire` / `failsafe`), and this PR takes care of adding the extra input based on a property `addQuarkusInputs`.

### Changes
- Dynamic dependencies added as extra input to the `build` and `test` goals
- README improved with some schema to detail the flow
- Clarify implementation by extracting Configuration and Utility methods in dedicated classes
- Add a prefix to log message for easier runtime troubleshooting

### CODE REVIEW

**Comment by jprinet** [comment]
FYI @aloubyansky @gsmet

File: quarkus-build-caching-extension/README.md [hunk 1]
```diff
@@ -105,6 +109,56 @@ The configuration file location can either be defined:
...
 
+## Implementation details
+
+The instructions required to make the Quarkus `build` goal cacheable are isolated to the [QuarkusBuildCache](./src/main/java/com/gradle/QuarkusBuildCache.java) class.
+
+### Quarkus configuration dump
+A key component of the caching mechanism is the Quarkus configuration dump file `.quarkus/quarkus-prod-config-dump`.
+This file is generated by the Quarkus `build` goal when the Maven property `quarkus.config-tracking.enabled` is `true`.
+It contains all the Quarkus properties used during the Quarkus `build` process.
+Some properties are discovered late in the `build` phase and can't be determined advance, thus the need for a first full execution to generate the file.
```
Comment (aloubyansky) [inline quarkus-build-caching-extension/README.md:120]:
```suggestion
Some properties are discovered late in the `build` phase and can't be determined in advance, thus the need for a first full execution to generate the file.
```

Comment (jprinet) [reply quarkus-build-caching-extension/README.md]: Thanks, I fixed it

File: quarkus-build-caching-extension/README.md [hunk 2]
```diff
@@ -1,52 +1,50 @@
...
-
+This project performs programmatic configuration of the [Develocity Build Cache](https://docs.gradle.com/enterprise/maven-extension/#using_the_build_cache) through a Maven extension. See [here](https://docs.gradle.com/enterprise/maven-extension/#custom_extension) for more details.
 This project is derived from [the CCUD Maven extension](https://github.com/gradle/common-custom-user-data-maven-extension).
 
-The instructions required to make the Quarkus build goal cacheable are isolated to the [QuarkusBuildCache](./src/main/java/com/gradle/QuarkusBuildCache.java) class.
-
-**It is important to notice that a native executable can be a very large file, and copying it from/to the local cache, or transferring it from/to the remote cache can be an expensive operation that has to be balanced with the caching benefit**
+*Note:*<br>
+A native executable can be a very large file, and copying it from/to the local cache, or transferring it from/to the remote cache can be an expensive operation that has to be balanced with the caching benefit
```
Comment (runningcode) [inline quarkus-build-caching-extension/README.md:8]:
```suggestion
A native executable can be a very large file. Copying it from/to the local cache, or transferring it from/to the remote cache can be an expensive operation that has to be balanced with the duration of the work being avoided.
```

File: quarkus-build-caching-extension/README.md [hunk 3]
```diff
@@ -1,52 +1,50 @@
...
-Only the *native*, *uber-jar*, *jar* and *legacy-jar* [packaging types](https://quarkus.io/guides/maven-tooling#quarkus-package-pkg-package-config_quarkus.package.type) can be made cacheable.
-
-The native packaging is cacheable only if the in-container build strategy (`quarkus.native.container-build=true`) is configured along with a fixed build image (`quarkus.native.builder-image`).
-This in-container build strategy means the build is as reproducible as possible. Even so, some timestamps and instruction ordering may be different even when built on the same system in the same environment.
-
-**Note:**
-*When the in-container build strategy is used as a fallback due to GraalVM requirements not met for instance, the caching feature will still be disabled. The recommendation is to explicitly set the in-container strategy (`quarkus.native.container-build=true`) to benefit from caching.* 
+**Note:**<br>
+- When the in-container build strategy is used as a fallback due to GraalVM requirements not met for instance, the caching feature will still be disabled. The recommendation is to explicitly set the in-container strategy (`quarkus.native.container-build=true`) to benefit from caching
```
Comment (runningcode) [inline quarkus-build-caching-extension/README.md:18]:
```suggestion
- When the in-container build strategy is used as a fallback the caching feature will be disabled. The fallback may happen due to GraalVM requirements not met. The recommendation is to explicitly set the in-container strategy (`quarkus.native.container-build=true`) to benefit from caching
```

File: quarkus-build-caching-extension/README.md [hunk 4]
```diff
@@ -77,7 +75,13 @@ Add the `track-prod-config-changes` execution to the `quarkus-maven-plugin` conf
 </plugin>
 ```
 
-The generated `.quarkus/quarkus-prod-config-dump` file must be checked in the project repository to allow a cache hit on a first build after cloning.
+## Improving the cache hit rate
+The file `.quarkus/quarkus-prod-config-dump` generated by the Quarkus build `goal` must be present to allow a cache hit.
+
+In order to increase the cache hit rate, it is possible to check-in the file in the source code repository.
+in continuous integration environment, an alternative is to have a cache restoring the file.
```
Comment (runningcode) [inline quarkus-build-caching-extension/README.md:82]: `an alternative is to have a cache restoring the file.`

The word cache here can be misleading. I assume you mean some sort of CI store/restore functionality? Not the build cache?

File: quarkus-build-caching-extension/README.md [hunk 5]
```diff
@@ -105,6 +109,56 @@ The configuration file location can either be defined:
 - as a maven property:
 `<gradle.quarkus.extension.config.file>.quarkus/extension-local.properties</gradle.quarkus.extension.config.file>`
 
+## Implementation details
+
+The instructions required to make the Quarkus `build` goal cacheable are isolated to the [QuarkusBuildCache](./src/main/java/com/gradle/QuarkusBuildCache.java) class.
```
Comment (runningcode) [inline quarkus-build-caching-extension/README.md:114]: by instructions here do you mean instructions for the consumer of the caching utility? Or do you mean the code to make the Quarkus build goal cacheable?

File: quarkus-build-caching-extension/README.md [hunk 6]
```diff
@@ -105,6 +109,56 @@ The configuration file location can either be defined:
...
+Let's illustrate the extension behavior with the following sequence of builds:
+
+**During the first build:**
+- `track-config-changes` does nothing as `.quarkus/quarkus-prod-config-dump` is absent
+- Quarkus configuration from current and previous build differ
+
+  => The `build` goal is not cacheable
+- `build` executes and creates `.quarkus/quarkus-prod-config-dump`
+
+![Run1](./doc/run-1.png)
```
Comment (runningcode) [inline quarkus-build-caching-extension/README.md:139]: nice to add these diagrams!

File: quarkus-build-caching-extension/src/main/java/com/gradle/QuarkusExtensionConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.util.Properties;
+
+final class QuarkusExtensionConfiguration {
+
+    // Environment variable key to disable caching
+    private static final String GRADLE_QUARKUS_KEY_CACHE_ENABLED = "GRADLE_QUARKUS_CACHE_ENABLED";
```
Comment (runningcode) [inline quarkus-build-caching-extension/src/main/java/com/gradle/QuarkusExtensionConfiguration.java:14]: Can we prefix these all with `DEVELOCITY` instead of `GRADLE`?

**Review by runningcode — COMMENTED** [review]
Gave it a review mostly on the documentation/readme.
Great work!

File: quarkus-build-caching-extension/README.md (hunk 4 above)
Comment (jprinet) [reply quarkus-build-caching-extension/README.md:82]: Correct, I'll update with
> in continuous integration environment, an alternative is to have the file restored (see [this option](https://github.com/marketplace/actions/cache#restoring-and-saving-cache-using-a-single-action) for GitHub actions as an example).

File: quarkus-build-caching-extension/README.md (hunk 5 above)
Comment (jprinet) [reply quarkus-build-caching-extension/README.md:114]: I meant the code, let me rephrase to clarify with
> The logic to make the Quarkus `build` goal cacheable is isolated to the [QuarkusBuildCache](./src/main/java/com/gradle/QuarkusBuildCache.java) class.

File: quarkus-build-caching-extension/src/main/java/com/gradle/QuarkusExtensionConfiguration.java (hunk 7 above)
Comment (jprinet) [reply quarkus-build-caching-extension/src/main/java/com/gradle/QuarkusExtensionConfiguration.java:14]: Sure

File: quarkus-build-caching-extension/README.md (hunk 5 above)
Comment (runningcode) [reply quarkus-build-caching-extension/README.md:114]: sounds good

File: quarkus-build-caching-extension/README.md (hunk 4 above)
Comment (runningcode) [reply quarkus-build-caching-extension/README.md:82]: i like this

**Review by runningcode — APPROVED** [review]
